sys/net/gnrc/network_layer/ndp2/gnrc_ndp2.c

```diff
@@ -0,0 +1,567 @@
+                                      uint8_t prefix_len,
+                                      uint32_t valid_ltime, uint32_t pref_ltime,
+                                      uint8_t flags, gnrc_pktsnip_t *next)
+{
+    assert(prefix != NULL);
+    assert(!ipv6_addr_is_link_local(prefix) && !ipv6_addr_is_multicast(prefix));
+    assert(prefix_len <= 128);
+    gnrc_pktsnip_t *pkt = gnrc_ndp2_opt_build(NDP_OPT_PI, sizeof(ndp_opt_pi_t),
+                                              next);
+
+    if (pkt != NULL) {
+        ndp_opt_pi_t *pi_opt = pkt->data;
+
+        pi_opt->prefix_len = prefix_len;
+        pi_opt->flags = (flags & NDP_OPT_PI_FLAGS_MASK);
+        pi_opt->valid_ltime = byteorder_htonl(valid_ltime);
+        pi_opt->pref_ltime = byteorder_htonl(pref_ltime);
+        pi_opt->resv.u32 = 0;
+        /* Bits beyond prefix_len MUST be 0 */
+        ipv6_addr_set_unspecified(&pi_opt->prefix);
+        ipv6_addr_init_prefix(&pi_opt->prefix, prefix, prefix_len);
+    }
+    return pkt;
+}
+
+gnrc_pktsnip_t *gnrc_ndp2_opt_mtu_build(uint32_t mtu, gnrc_pktsnip_t *next)
+{
+    gnrc_pktsnip_t *pkt = gnrc_ndp2_opt_build(NDP_OPT_MTU,
+                                              sizeof(ndp_opt_mtu_t), next);
+
+    if (pkt != NULL) {
+        ndp_opt_mtu_t *mtu_opt = pkt->data;
+
+        mtu_opt->resv.u16 = 0;
+        mtu_opt->mtu = byteorder_htonl(mtu);
+    }
+    return pkt;
+}
+
+static gnrc_pktsnip_t *_build_headers(gnrc_ipv6_netif_t *netif,
+                                      const ipv6_addr_t *src,
+                                      const ipv6_addr_t *dst,
+                                      gnrc_pktsnip_t *payload);
+static size_t _get_l2src(gnrc_ipv6_netif_t *netif, uint8_t *l2src,
+                         size_t l2src_maxlen);
+
+void gnrc_ndp2_nbr_sol_send(const ipv6_addr_t *tgt, gnrc_ipv6_netif_t *netif,
+                            const ipv6_addr_t *src, const ipv6_addr_t *dst)
+{
+    assert((tgt != NULL) && !ipv6_addr_is_multicast(tgt));
+    assert((netif != NULL) && (dst != NULL));
+    gnrc_pktsnip_t *hdr, *pkt = NULL;
+    /* cppcheck-suppress variableScope
+     * (reason: also used in MODULE_GNRC_SIXLOWPAN_ND compile path) */
+    uint8_t l2src[8];
+    /* cppcheck-suppress variableScope
+     * (reason: also used in MODULE_GNRC_SIXLOWPAN_ND compile path) */
+    size_t l2src_len = 0;
+
+    DEBUG("ndp2: send neighbor solicitation (iface: %" PRIkernel_pid ", "
+          "src: %s, ", netif->pid,
+          ipv6_addr_to_str(addr_str, (src != NULL) ? src : &ipv6_addr_unspecified,
+                           sizeof(addr_str)));
+    DEBUG("tgt: %s, ", ipv6_addr_to_str(addr_str, tgt, sizeof(addr_str)));
+    DEBUG("dst: %s)\n", ipv6_addr_to_str(addr_str, dst, sizeof(addr_str)));
+    /* check if there is a fitting source address to target */
+    if (src == NULL) {
+        src = gnrc_ipv6_netif_find_best_src_addr(netif->pid, tgt, false);
+    }
+
+    /* add SL2AO based on interface and source address */
```

Comment:
Yes, but I prefer it this way to distinguish link-local  (LL) from link-layer/layer 2 (L2)